EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
>1. changing the primary height source in flight sounds like a bad idea anyway

Yes, but we have to account for that.

>2. if not with sudden altitude change, the drone should converge to the commanded height. In your example the drone will keep an altitude error. Let's say the drone is flying at 2m altitude using baro. Then it moves at constant altitude above a 1.5 meter hill, now baro altitude is still 2m, but the range sensor returns a height of 0.5m. By switching to range sensor and updating _hgt_sensor_offset, the drone will stay 0.5m above ground, which is not the commanded height of 2m.

Yes, it stays at 0.5m. But that's what I would expect. I think it should use the altitude at the time the user switches.
@priseborough what do you think?